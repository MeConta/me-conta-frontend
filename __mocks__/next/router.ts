const router = {
  push: jest.fn()
}

export default router
