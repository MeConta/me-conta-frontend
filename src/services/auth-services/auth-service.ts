import { AxiosInstance } from 'axios'
import { UserType } from 'enums/user-type.enum'

type ResetSenhaForm = {
  hash: string
  senha: string
}

type LoginResponse = {
  token: string
  tipo: UserType
  nome: string
  refreshToken: string
}

export interface IAuthService {
  logout(): Promise<void>
  recuperarSenha?(email: string): Promise<void>
  resetarSenha?(form: ResetSenhaForm): Promise<void>
}

export class AuthService implements IAuthService {
  constructor(private readonly service: AxiosInstance) {}

  async logout() {
    await this.service.post('/auth/logout', {})
  }

  async recuperarSenha(email: string): Promise<void> {
    await this.service.post('/senha/recuperacao/', {
      email
    })
  }

  async resetarSenha(form: ResetSenhaForm): Promise<void> {
    await this.service.post('/senha/reset/', {
      hash: form.hash,
      senha: form.senha
    })
  }

  async refreshToken({
    refreshToken
  }: {
    refreshToken: string
  }): Promise<LoginResponse> {
    const response = await this.service.post('/auth/refresh', { refreshToken })
    return {
      refreshToken: response.data.refreshToken,
      token: response.data.token,
      nome: response.data.nome,
      tipo: response.data.tipo
    }
  }
}
