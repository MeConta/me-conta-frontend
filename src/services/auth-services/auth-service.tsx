import { AxiosInstance } from 'axios'
import { UserType } from 'enums/user-type.enum'
import {
  createContext,
  PropsWithChildren,
  useContext,
  useEffect,
  useState
} from 'react'
import { useLocalStorage } from '../../hooks/localstorage.hook'
import { parseJwtToObject } from 'utils/convertions/convertions-jwt'

type LoginForm = {
  email: string
  senha: string
}

type ResetSenhaForm = {
  hash: string
  senha: string
}

type LoginResponse = { token: string; tipo: UserType; nome: string }

export interface IAuthService {
  login(form: LoginForm): Promise<LoginResponse>
  logout(): Promise<void>
  recuperarSenha?(email: string): Promise<void>
  resetarSenha?(form: ResetSenhaForm): Promise<void>
}

export class AuthService implements IAuthService {
  constructor(private readonly service: AxiosInstance) {}
  async login(form: LoginForm): Promise<LoginResponse> {
    const response = await this.service.post('/auth/login/', {
      username: form.email,
      password: form.senha
    })

    //registrar no ssr

    return {
      token: response.data.token,
      tipo: parseInt(response.data.tipo, 10) as UserType,
      nome: response.data.nome
    }
  }

  async logout() {
    await this.service.post('/auth/logout', {})
  }

  async recuperarSenha(email: string): Promise<void> {
    await this.service.post('/senha/recuperacao/', {
      email
    })
  }

  async resetarSenha(form: ResetSenhaForm): Promise<void> {
    await this.service.post('/senha/reset/', {
      hash: form.hash,
      senha: form.senha
    })
  }
}

type SessionData = {
  nome: string
  tipo: string
  token: string
}

type AuthServiceProps = {
  authService: IAuthService
  isLoggedIn: boolean | null
  session: SessionData
  storeSessionData: (session: SessionData) => void
  clearSessionData: () => void
}

const LocalStorageAuthKeys: { [key: string]: string } = {
  TOKEN: 'token',
  NOME: 'nome',
  TIPO: 'tipo'
}

const AuthorizationContext = createContext<AuthServiceProps>(
  {} as AuthServiceProps
)

export const AuthorizationProvider = (
  props: PropsWithChildren<{ authService: IAuthService }>
) => {
  const [isLoggedIn, setIsLoggedIn] = useState<boolean | null>(null)
  const [token, setToken] = useLocalStorage(LocalStorageAuthKeys.TOKEN, '')
  const [nome, setNome] = useLocalStorage(LocalStorageAuthKeys.NOME, '')

  const decodedToken = parseJwtToObject(token)

  const [tipo, setTipo] = useState<string>(
    decodedToken?.roles?.[0]?.toString() || ''
  )

  useEffect(() => {
    if (decodedToken && decodedToken.roles) {
      setTipo(decodedToken.roles[0].toString())
    }
  }, [decodedToken])

  useEffect(() => {
    if (!token) {
      setIsLoggedIn(false)
    } else {
      if (
        !decodedToken ||
        (decodedToken && decodedToken.exp * 1000 < Date.now())
      ) {
        setIsLoggedIn(false)
      } else {
        setIsLoggedIn(true)
      }
    }
  }, [token, setIsLoggedIn])

  const storeSessionDataHandler = (session: SessionData) => {
    setToken(session.token)
    setNome(session.nome)
    setTipo(session.tipo)

    // store in the next session
  }

  const clearSessionDataHandler = () => {
    setToken('')
    setNome('')
    setTipo('')

    //remove from next session
  }

  const session: SessionData = {
    nome,
    tipo,
    token
  }

  return (
    <AuthorizationContext.Provider
      value={{
        authService: props.authService,
        isLoggedIn: isLoggedIn,
        session,
        storeSessionData: storeSessionDataHandler,
        clearSessionData: clearSessionDataHandler
      }}
    >
      {props.children}
    </AuthorizationContext.Provider>
  )
}

export function useAuthService(): AuthServiceProps {
  const context = useContext(AuthorizationContext)
  return context
}
