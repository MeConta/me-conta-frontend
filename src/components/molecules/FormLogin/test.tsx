import userEvent from '@testing-library/user-event'
import { UserType } from 'enums/user-type.enum'
import {
  fireEvent,
  render,
  RenderResult,
  screen,
  waitFor
} from 'utils/tests/helpers'
import { FormLogin } from '.'
import router from '../../../../__mocks__/next/router'
import { IAuthService } from '../../../services/auth-services/auth-service'

jest.mock('services/auth-services/auth-service', () => {
  const useAuthService = () => {
    return {
      storeSessionData: () => {},
      session: {
        nome: 'John Doe'
      }
    }
  }
  return { useAuthService }
})

const preencherFormularioParaSubmeter = async () => {
  const email = screen.getByRole('textbox', { name: 'E-mail' })
  const password = screen.getByRole('password', { name: 'Senha' })
  const submit = screen.getByRole('button')

  await userEvent.type(email, 'email@teste.com')
  await userEvent.type(password, 'S3nh@valid@')
  await fireEvent.click(submit)
}

describe('<FormLogin/>', () => {
  let renderResult: RenderResult
  const authServiceMock: IAuthService = {
    login: jest.fn()
  }
  const handleErrorMock = jest.fn()

  beforeEach(() => {
    renderResult = render(
      <FormLogin authService={authServiceMock} handleError={handleErrorMock} />
    )
  })

  it('deve renderizar todos os elementos do formulário', () => {
    expect(screen.getByRole('textbox', { name: 'E-mail' })).toBeInTheDocument()
    expect(screen.getByRole('password', { name: 'Senha' })).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /entrar/i })).toBeInTheDocument()

    expect(renderResult.container.parentElement).toMatchSnapshot()
  })

  it('deve exibir error de email inválido', async () => {
    const email = screen.getByRole('textbox', { name: 'E-mail' })
    const submit = screen.getByRole('button')

    await userEvent.type(email, 'meuemailcom')
    await fireEvent.click(submit)

    await waitFor(() => {
      expect(screen.getByText(/E-mail inválido/)).toBeInTheDocument()
    })
  })

  it('deve chamar o serviço de login', async () => {
    await preencherFormularioParaSubmeter()

    jest.spyOn(authServiceMock, 'login').mockImplementation(() => {
      return Promise.resolve({
        token: 'XPTO',
        tipo: UserType.ALUNO,
        nome: 'John'
      })
    })

    await waitFor(() => {
      expect(authServiceMock.login).toBeCalled()
    })
  })

  it('deve redirecionar para dashboard de aluno', async () => {
    await preencherFormularioParaSubmeter()

    jest
      .spyOn(authServiceMock, 'login')
      .mockImplementation(() =>
        Promise.resolve({ token: 'XPTO', tipo: UserType.ALUNO, nome: 'John' })
      )

    await waitFor(() => {
      expect(router.push).toBeCalledWith('/dashboard')
    })
  })

  it('deve redirecionar para dashboard de atendente', async () => {
    await preencherFormularioParaSubmeter()

    jest.spyOn(authServiceMock, 'login').mockImplementation(() =>
      Promise.resolve({
        token: 'XPTO',
        tipo: UserType.ATENDENTE,
        nome: 'John'
      })
    )

    await waitFor(() => {
      expect(router.push).toBeCalledWith('/dashboard-atendente')
    })
  })

  it('deve redirecionar para dashboard de supervisor', async () => {
    await preencherFormularioParaSubmeter()

    jest.spyOn(authServiceMock, 'login').mockImplementation(() =>
      Promise.resolve({
        token: 'XPTO',
        tipo: UserType.SUPERVISOR,
        nome: 'John'
      })
    )

    await waitFor(() => {
      expect(router.push).toBeCalledWith('/dashboard-supervisor')
    })
  })

  it('deve submeter formulário e chamar callback de error', async () => {
    await preencherFormularioParaSubmeter()

    jest
      .spyOn(authServiceMock, 'login')
      .mockImplementation(() => Promise.reject())

    await waitFor(() => {
      expect(authServiceMock.login).toBeCalled()
      expect(handleErrorMock).toBeCalled()
    })
  })
})
