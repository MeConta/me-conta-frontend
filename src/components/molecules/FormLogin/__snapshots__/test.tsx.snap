// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormLogin/> should render a login form  1`] = `
.c6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  color: #FFFFFF;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  text-align: center;
  min-height: 5rem;
  font-size: 2.2rem;
  padding: 0.8rem 4.8rem;
  background: #458FF6;
  border-radius: 0.7rem;
}

.c6:hover {
  background-color: #0545AB;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
}

.c5 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  margin-bottom: 1.6rem;
}

.c1 {
  font-size: 1.5rem;
  font-weight: 300;
  color: #848A8C;
  margin-left: 0.8rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.7rem;
  padding: 0.8rem 1.6rem;
  margin-top: 0.8rem;
  border: 0.1rem solid;
  border-color: #848A8C;
}

.c2:focus-within {
  box-shadow: 0 0 0.5rem #0545AB;
}

.c3 {
  color: #000000;
  font-family: Mulish,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.5rem;
  padding: 0.5rem 0;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c3:disabled {
  cursor: not-allowed;
  color: #828282;
}

.c3:disabled::-webkit-input-placeholder {
  color: currentColor;
}

.c3:disabled::-moz-placeholder {
  color: currentColor;
}

.c3:disabled:-ms-input-placeholder {
  color: currentColor;
}

.c3:disabled::placeholder {
  color: currentColor;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #828282;
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
}

.c4 > svg {
  width: 2rem;
  height: 100%;
}

<body>
  <div>
    <form>
      <h2>
        Fa√ßa seu login e comece seu atendimento!
      </h2>
      <div
        class="c0"
      >
        <label
          class="c1"
          for="email"
        >
          E-mail
        </label>
        <div
          class="c2"
        >
          <input
            class="c3"
            id="email"
            name="email"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="c0"
      >
        <label
          class="c1"
          for="password"
        >
          Password
        </label>
        <div
          class="c2"
        >
          <div
            class="c4"
            data-testid="icon"
          >
            <svg
              aria-hidden="true"
              aria-label="Mostrar Senha"
              class="c5"
              fill="currentColor"
              focusable="false"
              viewBox="0 0 16 16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10.5 8a2.5 2.5 0 1 1-5 0 2.5 2.5 0 0 1 5 0z"
              />
              <path
                d="M0 8s3-5.5 8-5.5S16 8 16 8s-3 5.5-8 5.5S0 8 0 8zm8 3.5a3.5 3.5 0 1 0 0-7 3.5 3.5 0 0 0 0 7z"
              />
            </svg>
          </div>
          <input
            class="c3"
            id="password"
            name="password"
            role="password"
            type="password"
            value=""
          />
        </div>
      </div>
      <a
        href="/forgot-password"
      >
        Esqueceu a senha?
      </a>
      <button
        class="c6"
        color="primary"
        radius="square"
        type="submit"
      >
        ENTRAR
      </button>
      <p>
        Quer fazer parte do Me Conta?
      </p>
    </form>
  </div>
</body>
`;
